pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -187,40 +212,119 @@ func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(namespacedName string)
 	return nil
 }
 
+// Find IP Pools annotated to StatefulSet via direct annotation or namespace annotation
+func (c *AntreaIPAMController) getIPPoolsForStatefulSet(ss *appsv1.StatefulSet) []string {
+
+	// Inspect pool annotation for the Pods
+	// In order to avoid extra API call in IPAM driver, IPAM annotations are defined
+	// on Pods rather than on StatefulSet
+	annotations, exists := ss.Spec.Template.Annotations[annotation.AntreaIPAMAnnotationKey]
+	if exists {
+		// Stateful Set Pod is annotated with dedicated IP pool
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	// Inspect Namespace
+	namespace, err := c.namespaceLister.Get(ss.Namespace)
+	if err != nil {
+		// Should never happen
+		klog.Errorf("Namespace %s not found for StatefulSet %s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	annotations, exists = namespace.Annotations[annotation.AntreaIPAMAnnotationKey]
+	if exists {
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the namespace. If such IP Pool is found, preallocate IPs for the StatefulSet.
+// This function returns error if pool is not found, or allocation fails.
+func (c *AntreaIPAMController) preallocateIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing create notification", "Namespace", ss.Namespace, "StatefulSet", ss.Name)
+
+	ipPools := c.getIPPoolsForStatefulSet(ss)
+
+	if ipPools == nil {
+		// nothing to preallocate
+		return nil
+	}
+
+	// Only one pool is supported for now. Dual stack support coming in future.
+	ipPoolName := ipPools[0]
```

Comment:
I'm not sure if we should just select the first pool or raise an error.
We don't know what is the user intention. Perhaps they made a mistake in specification and wanted to use the pool that we decided to ignore.

Since we only support a single pool I tend to think it might be ok to raise an error if multiple pools are specified. Removing the check in future releases should not lead to bw-incompatible changes.

Reply:
I've added error handling for multiple pools case